ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdRecalculationService.java

```diff
@@ -114,16 +128,35 @@ public void recalculatePersistedInstanceId(String collectionName) {
 					collectionName);
 			return;
 		}
-		int numberOfPartitions = determineNumberOfPartitions();
-		recalculatePersistedInstanceId(collectionName, numberOfPartitions);
+		recalculatePersistedInstanceId(collectionName, getNumberOfPartitionsToCalculate());
 	}
 
-	private void recalculatePersistedInstanceId(String collectionName, int numberOfPartitions) {
-		log.info("Recalculating persisted instance id for collection {} with {} number of partitions", collectionName, numberOfPartitions);
+	private Predicate<IndexInfo> isInstanceIdIndex() {
+		return indexInfo -> indexInfo.getIndexFields().size() == 1 && indexInfo.getIndexFields().get(0).getKey().startsWith(DOCUMENT_INSTANCE_ID);
+	}
+
+	private void recalculatePersistedInstanceId(String collectionName, Set<Integer> numberOfPartitionsSet) {
+		log.info("Recalculating persisted instance id for collection {} with {} number of partitions",
+				collectionName,
+				numberOfPartitionsSet.stream().sorted().collect(toList())
+		);
 		DocumentCollection collection = spaceMirror.getDocumentDb().getCollection(collectionName);
-		boolean indexDropped = collection.getIndexes()
-				.filter(index -> index.isIndexForFields(Set.of(DOCUMENT_INSTANCE_ID)))
-				.filter(not(isIndexForNumberOfPartitions(numberOfPartitions)))
+
+		Set<String> fieldNamesToCalculate = numberOfPartitionsSet.stream()
+				.map(PersistedInstanceIdUtil::getInstanceIdFieldName)
+				.collect(toSet());
+
+		List<IndexInfo> allInstanceIdIndexes = collection.getIndexes()
+				.filter(isInstanceIdIndex())
+				.collect(toList());
+
+		Set<String> noLongerNeededFields = allInstanceIdIndexes.stream()
+				.map(index -> index.getIndexFields().get(0).getKey())
+				.filter(fieldName -> !fieldNamesToCalculate.contains(fieldName))
+				.collect(toSet());
```

Comment:
If the job is cancelled before being finished, `noLongerNeededFields` might not be removed for all objects, and will not be removed by running the job again as these indices are dropped at the start of the job. This should not be a large problem, as it is only unused data and will be removed if the object is updated.

Discovering actually existing `_instanceId` by inspecting all objects in mongo would be one solution to this.